File: backend/api/routes/favorites.py
"""
Favorites API routes
"""
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import List, Optional
from pydantic import BaseModel

from database.session import get_db
from database.models.user import Favorite, User
from database.models.classroom import Classroom
from api.routes.auth import session_store

router = APIRouter(prefix="/favorites", tags=["favorites"])


def get_current_user_id(token: Optional[str] = Query(None)) -> str:
    """Get current user ID from token"""
    if not token:
        raise HTTPException(status_code=401, detail="Authentication required")
    
    if token not in session_store:
        raise HTTPException(status_code=401, detail="Invalid or expired token")
    
    user_id = session_store[token].get("user_id")
    if not user_id:
        raise HTTPException(status_code=401, detail="Invalid session")
    
    return user_id


class FavoriteResponse(BaseModel):
    """Favorite response model"""
    id: str
    classroom_id: str
    classroom: dict
    created_at: str
    
    class Config:
        from_attributes = True


@router.get("/", response_model=List[FavoriteResponse])
async def get_favorites(
    token: Optional[str] = Query(None),
    db: Session = Depends(get_db)
):
    """Get user's favorite classrooms"""
    user_id = get_current_user_id(token)
    
    favorites = db.query(Favorite).filter(Favorite.user_id == user_id).all()
    
    result = []
    for fav in favorites:
        classroom = db.query(Classroom).filter(Classroom.id == fav.classroom_id).first()
        if classroom:
            result.append({
                "id": fav.id,
                "classroom_id": fav.classroom_id,
                "classroom": {
                    "id": classroom.id,
                    "room_number": classroom.room_number,
                    "building_id": classroom.building_id,
                    "faculty": classroom.faculty,
                    "floor": classroom.floor,
                    "capacity": classroom.capacity,
                    "has_projector": classroom.has_projector,
                    "has_wifi": classroom.has_wifi,
                    "has_power_outlets": classroom.has_power_outlets,
                },
                "created_at": str(fav.created_at),
            })
    
    return result


@router.post("/{classroom_id}")
async def add_favorite(
    classroom_id: str,
    token: Optional[str] = Query(None),
    db: Session = Depends(get_db)
):
    """Add classroom to favorites"""
    user_id = get_current_user_id(token)
    
    # Check if classroom exists
    classroom = db.query(Classroom).filter(Classroom.id == classroom_id).first()
    if not classroom:
        raise HTTPException(status_code=404, detail="Classroom not found")
    
    # Check if already favorited
    existing = db.query(Favorite).filter(
        Favorite.user_id == user_id,
        Favorite.classroom_id == classroom_id
    ).first()
    
    if existing:
        raise HTTPException(status_code=400, detail="Already in favorites")
    
    # Create favorite
    favorite = Favorite(
        id=f"fav_{user_id}_{classroom_id}",
        user_id=user_id,
        classroom_id=classroom_id,
    )
    db.add(favorite)
    db.commit()
    db.refresh(favorite)
    
    return {"message": "Added to favorites", "id": favorite.id}


@router.delete("/{classroom_id}")
async def remove_favorite(
    classroom_id: str,
    token: Optional[str] = Query(None),
    db: Session = Depends(get_db)
):
    """Remove classroom from favorites"""
    user_id = get_current_user_id(token)
    
    favorite = db.query(Favorite).filter(
        Favorite.user_id == user_id,
        Favorite.classroom_id == classroom_id
    ).first()
    
    if not favorite:
        raise HTTPException(status_code=404, detail="Favorite not found")
    
    db.delete(favorite)
    db.commit()
    
    return {"message": "Removed from favorites"}


@router.get("/check/{classroom_id}")
async def check_favorite(
    classroom_id: str,
    token: Optional[str] = Query(None),
    db: Session = Depends(get_db)
):
    """Check if classroom is in favorites"""
    try:
        user_id = get_current_user_id(token)
        favorite = db.query(Favorite).filter(
            Favorite.user_id == user_id,
            Favorite.classroom_id == classroom_id
        ).first()
        return {"is_favorite": favorite is not None}
    except HTTPException:
        return {"is_favorite": False}

File: client/lib/favorites.ts
/**
 * Favorites API client
 */
// In Vercel/production, use relative paths (same domain)
// In development, use localhost:8000
const API_BASE_URL = import.meta.env.VITE_API_URL || 
  (import.meta.env.PROD ? '' : 'http://localhost:8000');

export interface Favorite {
  id: string;
  classroom_id: string;
  classroom: {
    id: string;
    room_number: string;
    building_id: string;
    faculty: string;
    floor: number;
    capacity: number;
    has_projector: boolean;
    has_wifi: boolean;
    has_power_outlets: boolean;
  };
  created_at: string;
}

function getToken(): string | null {
  return localStorage.getItem('auth_token');
}

/**
 * Get user's favorites
 */
export async function getFavorites(): Promise<Favorite[]> {
  const token = getToken();
  if (!token) {
    throw new Error('Authentication required');
  }
  
  const response = await fetch(`${API_BASE_URL}/api/v1/favorites/?token=${token}`);
  if (!response.ok) {
    throw new Error('Failed to get favorites');
  }
  return response.json();
}

/**
 * Add classroom to favorites
 */
export async function addFavorite(classroomId: string): Promise<void> {
  const token = getToken();
  if (!token) {
    throw new Error('Authentication required');
  }
  
  const response = await fetch(
    `${API_BASE_URL}/api/v1/favorites/${classroomId}?token=${token}`,
    { method: 'POST' }
  );
  if (!response.ok) {
    throw new Error('Failed to add favorite');
  }
}

/**
 * Remove classroom from favorites
 */
export async function removeFavorite(classroomId: string): Promise<void> {
  const token = getToken();
  if (!token) {
    throw new Error('Authentication required');
  }
  
  const response = await fetch(
    `${API_BASE_URL}/api/v1/favorites/${classroomId}?token=${token}`,
    { method: 'DELETE' }
  );
  if (!response.ok) {
    throw new Error('Failed to remove favorite');
  }
}

/**
 * Check if classroom is in favorites
 */
export async function checkFavorite(classroomId: string): Promise<boolean> {
  const token = getToken();
  if (!token) {
    return false;
  }
  
  try {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/favorites/check/${classroomId}?token=${token}`
    );
    if (!response.ok) {
      return false;
    }
    const data = await response.json();
    return data.is_favorite;
  } catch {
    return false;
  }
}

